acra/guides/advanced-integrations/client-side-integration-with-acra-connector.md

```diff
@@ -0,0 +1,224 @@
+
+#### Reads
+
+Read requests are the same as the regular reads, but their remote address is AcraConnector instead of your database listener. A typical PostgreSQL connection address would look like this: `postgresql://user:password@127.0.0.1:9494/db_name`.
+
+> Note for PostgreSQL: In the current layout, you aren't required to use SSL when trying to connect to the database. Transport protection for sensitive (decrypted) data is provided between AcraServer and AcraConnector via [Secure Session]({{<ref "themis/crypto-theory/cryptosystems/secure-session.md" >}}). However, you can setup using SSL connection, too.
+
+### Client-side with Zones
+
+Description of the client-side with Zones for PostgreSQL.
+
+Zone-based encryption is the best way to cryptographically compartmentalise the data that comes from different sources, following the user's choice. Cryptographically, the Zones are a mechanism for modifying the AcraServer key according to the Zone Id and for mapping Zone's private keys during the decryption. In theory, you could use a new key for every record. However, it would introduce an unwanted overhead on AcraServer's performance.
+
+#### Writing
+
+To write AcraStruct with Zone keys, you need to:
+
+* generate the Zone keys on AcraServer,
+* acquire Zone public key and Zone Id from AcraServer,
+* encrypt data using this key as AcraServer key in a standard AcraStruct process.
+
+There are two ways you can generate the keys:
+
+- First, make sure that you already generated master key and exported it as ACRA_MASTER_KEY environmental variable, or generate it like this:
+```
+export ACRA_MASTER_KEY=$(echo -n "My_Very_Long_Key_Phrase_ge_32_chars" | base64)
+```
+Read more about different types of keys used in Acra on [Key Management]({{< ref "acra/acra-in-depth/cryptography-and-key-management/#-INVALID" >}}) page.
+
+- Then build `acra-addzone` utility in AcraServer from Acra repository:
+
+```
+go build acra-addzone
+./acra-addzone
+``` 
+When you run it, you should get JSON object that looks something like this:
+
+```
+{"id":"DDDDDDDDQHpbUSOgYTzqCktp","public_key":"VUVDMgAAAC3yMBGsAmK/wBXZkL8iBv/C+7hqoQtSZpYoi4fZYMafkJbWe2dL"}
+```
+
+By decoding base64-wrapped public key, you get the binary Zone public key, which can be used for generating AcraStructs.
+
+- AcraConnector locally provides HTTP API to add Zones and return public keys on `127.0.0.1:9191/getNewZone`. You can change the port via `--incoming_connection_api_port` argument.
+
+Both approaches provide identical results.
+
+#### Reading
+
+Reading with Zones takes place just as it does without Zones. However, to point AcraServer to the exact Zone key, it has to use, you need to structure your query in such a way that AcraStruct is preceded by the corresponding Zone Id in its answer. For example, you may want to store the Zone Id in the preceding column and execute the query in the following way:
```

Comment:
> Reading with Zones takes place just as it does without Zones.

Reading with zones is just like reading without zones but with zones.

A samurai without a sword is like a samurai with a sword but without a sword.

😄

I think you need to make it clear and prominent that if you are using zones then you *will* need to adjust your application queries for read path as well, as opposed to non-zoned mode where only the write path needs application-level changes while the read path needs only configuration changes.

Reply:
re-phrased to `Retrieving sensitive data using zones a bit different. To retrieve sensitive data you need to structure your ....`